moonraker/plugins/history.py

```diff
@@ -0,0 +1,253 @@
+
+    async def _init_ready(self):
+        klippy_apis = self.server.lookup_plugin('klippy_apis')
+        self.print_stats = {}
+
+        try:
+            result = await klippy_apis.query_objects({'print_stats': None})
+        except self.server.error as e:
+            logging.info(f"Error getting print_stats: {e}")
+        self.print_stats = result.get("print_stats", {})
+
+        sub = {"print_stats": None}
+        try:
+            status = await klippy_apis.subscribe_objects(sub)
+        except self.server.error as e:
+            logging.info(f"Error subscribing to print_stats")
+
+    async def _handle_job_delete(self, web_request):
+        args = web_request.get_args()
+        if "all" in args:
+            deljobs = []
+            jobs = list(self.jobs)
+            for job in jobs:
+                self.delete_job(job)
+                deljobs.append(job)
+            return deljobs
+        if "id" not in args:
+            raise self.server.error("No ID to delete")
+        if args['id'] not in self.jobs:
+            raise self.server.error(f"Invalid job id: {args['id']}")
+
+        logging.debug("Jobs %s: %s" % (args['id'], self.jobs))
+        return self.delete_job(args['id'])
+
+    async def _handle_jobs_list(self, web_request):
+        args = web_request.get_args()
+        if "id" in args:
+            if args['id'] not in self.jobs:
+                raise self.server.error(f"Invalid job id: {args['id']}")
+            return {args['id']: self.get_job(args['id']).get_stats()}
+
+        before = None if 'before' not in args else int(args['before'])
+        since = None if 'since' not in args else int(args['since'])
+        limit = int(args['limit']) if "limit" in args else 50
+        start = int(args['start']) if "start" in args else 0
+        if start > (len(self.jobs)-1) or len(self.jobs) == 0:
+            return {"count": len(self.jobs), "prints": {}}
+
+        jobs = {}
+        i = 0
+        start_num = 0
+        end_num = len(self.jobs)
+        for id in list(self.jobs):
+            if since != None and since > self.get_job(id).get('start_time'):
+                start_num += 1
+                continue
+            if before != None and before < self.get_job(id).get('end_time'):
+                end_num -= 1
+                continue
+            if limit != 0 and i >= limit:
+                continue
+            if start != 0:
+                start -= 1
+                continue
+            jobs[id] = self.get_job(id).get_stats()
+            i += 1
+
+        return {"count": end_num - start_num, "prints": jobs}
+
+    async def _status_update(self, data):
+        if "print_stats" in data:
+            ps = data['print_stats']
+
+            if "filename" in ps:
+                file_store = self.server.lookup_plugin('file_manager')
+                if ps['filename'] != "":
+                    self.file_metadata = file_store.get_file_metadata(
+                        ps['filename'])
```

Comment:
Since we have direct access to the metadata through the database it isn't necessary to use the file manager to retrieve it.  We could do something like:
```python
self.file_metadata = self.gcdb.get(ps['filename'])
```